File: sdk/cs/src/FoundryLocalManager.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------
namespace Microsoft.AI.Foundry.Local;

using System;
using System.Text.Json;
using System.Threading.Tasks;

using Microsoft.AI.Foundry.Local.Detail;
using Microsoft.Extensions.Logging;

public class FoundryLocalManager : IDisposable
{
    private static FoundryLocalManager? instance;
    private static readonly AsyncLock asyncLock = new();

    internal static readonly string AssemblyVersion =
        typeof(FoundryLocalManager).Assembly.GetName().Version?.ToString() ?? "unknown";

    private readonly Configuration _config;
    private CoreInterop _coreInterop = default!;
    private Catalog _catalog = default!;
    private ModelLoadManager _modelManager = default!;
    private readonly AsyncLock _lock = new();
    private bool _disposed;
    private readonly ILogger _logger;

    internal Configuration Configuration => _config;
    internal ILogger Logger => _logger;
    internal ICoreInterop CoreInterop => _coreInterop!; // always valid once the instance is created

    public static bool IsInitialized => instance != null;
    public static FoundryLocalManager Instance => instance ??
        throw new FoundryLocalException("FoundryLocalManager has not been created. Call CreateAsync first.");

    /// <summary>
    /// Bound Urls if the web service has been started. Null otherwise.
    /// See <see cref="StartWebServiceAsync"/>.
    /// </summary>
    public string[]? Urls { get; private set; }

    /// <summary>
    /// Create the <see cref="FoundryLocalManager"/> singleton instance.
    /// </summary>
    /// <param name="configuration">Configuration to use.</param>
    /// <param name="logger">Application logger to use.
    /// Use Microsoft.Extensions.Logging.NullLogger.Instance if you wish to ignore log output from the SDK.
    /// </param>
    /// <param name="ct">Optional cancellation token for the initialization.</param>
    /// <returns>Task creating the instance.</returns>
    /// <exception cref="FoundryLocalException"></exception>
    public static async Task CreateAsync(Configuration configuration, ILogger logger,
                                         CancellationToken? ct = null)
    {
        using var disposable = await asyncLock.LockAsync().ConfigureAwait(false);

        if (instance != null)
        {
            // throw as we're not going to use the provided configuration in case it differs from the original.
            throw new FoundryLocalException("FoundryLocalManager has already been created.", logger);
        }

        FoundryLocalManager? manager = null;
        try
        {
            // use a local variable to ensure fully initialized before assigning to static instance.
            manager = new FoundryLocalManager(configuration, logger);
            await manager.InitializeAsync(ct).ConfigureAwait(false);

            // there is no previous as we only get here if instance is null.
            // ownership is transferred to the static instance.
#pragma warning disable IDISP003 // Dispose previous before re-assigning
            instance = manager;
            manager = null;
#pragma warning restore IDISP003
        }
        catch (Exception ex)
        {
            manager?.Dispose();

            if (ex is FoundryLocalException or OperationCanceledException)
            {
                throw;
            }

            // log and throw as FoundryLocalException
            throw new FoundryLocalException("Error during initialization.", ex, logger);
        }
    }

    /// <summary>
    /// Get the model catalog instance.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>The model catalog.</returns>
    /// <remarks>
    /// The catalog is populated on first use and returns models based on currently available execution providers.
    /// To ensure all hardware-accelerated models are listed, call <see cref="DownloadAndRegisterEpsAsync"/> first to
    /// register execution providers, then access the catalog.
    /// </remarks>
    public async Task<ICatalog> GetCatalogAsync(CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => GetCatalogImplAsync(ct),
                                                     "Error getting Catalog.", _logger).ConfigureAwait(false);
    }

    /// <summary>
    /// Start the optional web service. This will provide an OpenAI-compatible REST endpoint that supports
    ///   /v1/chat_completions
    ///   /v1/models to list downloaded models
    ///   /v1/models/{model_id} to get model details
    ///
    /// <see cref="Urls"/> is populated with the actual bound Urls after startup.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Task starting the web service.</returns>
    public async Task StartWebServiceAsync(CancellationToken? ct = null)
    {
        await Utils.CallWithExceptionHandling(() => StartWebServiceImplAsync(ct),
                                              "Error starting web service.", _logger).ConfigureAwait(false);
    }

    /// <summary>
    /// Stops the web service if started.
    /// </summary>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Task stopping the web service.</returns>
    public async Task StopWebServiceAsync(CancellationToken? ct = null)
    {
        await Utils.CallWithExceptionHandling(() => StopWebServiceImplAsync(ct),
                                              "Error stopping web service.", _logger).ConfigureAwait(false);
    }

    /// <summary>
    /// Discovers all available execution provider bootstrappers.
    /// Returns metadata about each EP including whether it is already registered.
    /// </summary>
    /// <returns>Array of EP bootstrapper info describing available EPs.</returns>
    public EpInfo[] DiscoverEps()
    {
        return Utils.CallWithExceptionHandling(DiscoverEpsImpl,
                                               "Error discovering execution providers.", _logger);
    }

    /// <summary>
    /// Downloads and registers execution providers. This is a blocking call that completes when all
    /// requested EPs have been processed.
    /// </summary>
    /// <param name="names">
    /// Optional subset of EP bootstrapper names to download (as returned by <see cref="DiscoverEps"/>).
    /// If null or empty, all discoverable EPs are downloaded.
    /// </param>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Result describing which EPs succeeded and which failed.</returns>
    /// <remarks>
    /// Catalog and model requests use whatever EPs are currently registered and do not block on EP downloads.
    /// After downloading new EPs, re-fetch the model catalog to include models requiring the newly registered EPs.
    /// </remarks>
    public async Task<EpDownloadResult> DownloadAndRegisterEpsAsync(IEnumerable<string>? names = null,
                                                          CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => DownloadAndRegisterEpsImplAsync(names, null, ct),
                                                     "Error downloading execution providers.", _logger)
                                                    .ConfigureAwait(false);
    }

    /// <summary>
    /// Downloads and registers execution providers with per-EP progress reporting.
    /// </summary>
    public async Task DownloadAndRegisterEpsAsync(IEnumerable<string>? names,
                                                  Action<string, double> progressCallback,
                                                  CancellationToken? ct = null)
    {
        await Utils.CallWithExceptionHandling(() => DownloadAndRegisterEpsImplAsync(names, progressCallback, ct),
                                              "Error downloading execution providers.", _logger)
                                             .ConfigureAwait(false);
    }

    private FoundryLocalManager(Configuration configuration, ILogger logger)
    {
        _config = configuration ?? throw new ArgumentNullException(nameof(configuration));
        _logger = logger;
    }

    private async Task InitializeAsync(CancellationToken? ct = null)
    {
        _config.Validate();
        _coreInterop = new CoreInterop(_config, _logger);

#pragma warning disable IDISP003 // Dispose previous before re-assigning. Always null when this is called.
        _modelManager = new ModelLoadManager(_config.Web?.ExternalUrl, _coreInterop, _logger);
#pragma warning restore IDISP003

        if (_config.ModelCacheDir != null)
        {
            CoreInteropRequest? input = null;
            var result = await _coreInterop!.ExecuteCommandAsync("get_cache_directory", input, ct)
                                            .ConfigureAwait(false);
            if (result.Error != null)
            {
                throw new FoundryLocalException($"Error getting current model cache directory: {result.Error}",
                                                _logger);
            }

            var curCacheDir = result.Data!;
            if (curCacheDir != _config.ModelCacheDir)
            {
                var request = new CoreInteropRequest
                {
                    Params = new Dictionary<string, string> { { "Directory", _config.ModelCacheDir } }
                };

                result = await _coreInterop!.ExecuteCommandAsync("set_cache_directory", request, ct)
                                            .ConfigureAwait(false);
                if (result.Error != null)
                {
                    throw new FoundryLocalException(
                        $"Error setting model cache directory to '{_config.ModelCacheDir}': {result.Error}", _logger);
                }
            }
        }

        return;
    }

    private EpInfo[] DiscoverEpsImpl()
    {
        var result = _coreInterop!.ExecuteCommand("discover_eps");
        if (result.Error != null)
        {
            throw new FoundryLocalException($"Error discovering execution providers: {result.Error}", _logger);
        }

        return JsonSerializer.Deserialize(result.Data!, JsonSerializationContext.Default.EpInfoArray)
            ?? Array.Empty<EpInfo>();
    }

    private async Task<ICatalog> GetCatalogImplAsync(CancellationToken? ct = null)
    {
        // create on first use
        if (_catalog == null)
        {
            using var disposable = await _lock.LockAsync().ConfigureAwait(false);
            if (_catalog == null)
            {
                _catalog = await Catalog.CreateAsync(_modelManager!, _coreInterop!, _logger, ct).ConfigureAwait(false);
            }
        }

        return _catalog;
    }

    private async Task StartWebServiceImplAsync(CancellationToken? ct = null)
    {
        if (_config?.Web?.Urls == null)
        {
            throw new FoundryLocalException("Web service configuration was not provided.", _logger);
        }

        using var disposable = await asyncLock.LockAsync().ConfigureAwait(false);

        CoreInteropRequest? input = null;
        var result = await _coreInterop!.ExecuteCommandAsync("start_service", input, ct).ConfigureAwait(false);
        if (result.Error != null)
        {
            throw new FoundryLocalException($"Error starting web service: {result.Error}", _logger);
        }

        var typeInfo = JsonSerializationContext.Default.StringArray;
        var boundUrls = JsonSerializer.Deserialize(result.Data!, typeInfo);
        if (boundUrls == null || boundUrls.Length == 0)
        {
            throw new FoundryLocalException("Failed to get bound URLs from web service start response.", _logger);
        }

        Urls = boundUrls;
    }

    private async Task StopWebServiceImplAsync(CancellationToken? ct = null)
    {
        if (_config?.Web?.Urls == null)
        {
            throw new FoundryLocalException("Web service configuration was not provided.", _logger);
        }

        using var disposable = await asyncLock.LockAsync().ConfigureAwait(false);

        CoreInteropRequest? input = null;
        var result = await _coreInterop!.ExecuteCommandAsync("stop_service", input, ct).ConfigureAwait(false);
        if (result.Error != null)
        {
            throw new FoundryLocalException($"Error stopping web service: {result.Error}", _logger);
        }

        // Should we clear these even if there's an error response?
        // Service is probably in a bad state or was not running.
        Urls = null;
    }

    private async Task<EpDownloadResult> DownloadAndRegisterEpsImplAsync(IEnumerable<string>? names = null,
                                                                Action<string, double>? progressCallback = null,
                                                                CancellationToken? ct = null)
    {
        using var disposable = await asyncLock.LockAsync().ConfigureAwait(false);

        CoreInteropRequest? input = null;
        if (names != null)
        {
            var namesList = string.Join(",", names);
            if (!string.IsNullOrEmpty(namesList))
            {
                input = new CoreInteropRequest
                {
                    Params = new Dictionary<string, string> { { "Names", namesList } }
                };
            }
        }

        ICoreInterop.Response result;

        if (progressCallback != null)
        {
            var callback = new ICoreInterop.CallbackFn(progressString =>
            {
                var sepIndex = progressString.IndexOf('|');
                if (sepIndex >= 0)
                {
                    var name = progressString[..sepIndex];
                    if (double.TryParse(progressString[(sepIndex + 1)..],
                                        System.Globalization.NumberStyles.Float,
                                        System.Globalization.CultureInfo.InvariantCulture,
                                        out var percent))
                    {
                        progressCallback(string.IsNullOrEmpty(name) ? "" : name, percent);
                    }
                }
            });

            result = await _coreInterop!.ExecuteCommandWithCallbackAsync("download_and_register_eps", input,
                                                                         callback, ct).ConfigureAwait(false);
        }
        else
        {
            result = await _coreInterop!.ExecuteCommandAsync("download_and_register_eps", input, ct).ConfigureAwait(false);
        }

        if (result.Error != null)
        {
            throw new FoundryLocalException($"Error downloading execution providers: {result.Error}", _logger);
        }

        EpDownloadResult epResult;

        if (!string.IsNullOrEmpty(result.Data))
        {
            epResult = JsonSerializer.Deserialize(result.Data!, JsonSerializationContext.Default.EpDownloadResult)
                ?? throw new FoundryLocalException("Failed to deserialize EP download result.", _logger);
        }
        else
        {
            epResult = new EpDownloadResult { Success = true, Status = "Completed", RegisteredEps = [], FailedEps = [] };
        }

        // Invalidate the catalog cache if any EP was newly registered so the next access
        // re-fetches models with the updated set of available EPs.
        if ((epResult.Success || epResult.RegisteredEps.Length > 0) && _catalog != null)
        {
            _catalog.InvalidateCache();
        }

        return epResult;
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (disposing)
            {
                if (Urls != null)
                {
                    // best effort stop
                    try
                    {
                        StopWebServiceImplAsync().GetAwaiter().GetResult();
                    }
                    catch (Exception ex)
                    {
                        _logger.LogWarning(ex, "Error stopping web service during Dispose.");
                    }
                }

                _catalog?.Dispose();
                _modelManager?.Dispose();
                _lock.Dispose();
            }

            _disposed = true;
        }
    }

    public void Dispose()
    {
        // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
        Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }
}

File: sdk/python/src/foundry_local_manager.py
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------

from __future__ import annotations

import asyncio
import json
import logging
import threading

from typing import Callable, List, Optional

from pydantic import TypeAdapter

from .catalog import Catalog
from .configuration import Configuration
from .ep_types import EpDownloadResult, EpInfo
from .logging_helper import set_default_logger_severity
from .detail.core_interop import CoreInterop, InteropRequest
from .detail.model_load_manager import ModelLoadManager
from .exception import FoundryLocalException

logger = logging.getLogger(__name__)


class FoundryLocalManager:
    """Singleton manager for Foundry Local SDK operations.

    Call ``await FoundryLocalManager.initialize(config)`` once at startup, then
    access the singleton via ``FoundryLocalManager.instance``.

    Attributes:
        instance: The singleton ``FoundryLocalManager`` instance (set after ``initialize``).
        catalog: The model ``Catalog`` for discovering and managing models.
        urls: Bound URL(s) after ``start_web_service()`` is called, or ``None``.
    """

    _lock = threading.Lock()
    instance: FoundryLocalManager = None

    @staticmethod
    async def initialize(config: Configuration):
        """Initialize the Foundry Local SDK with the given configuration.

        This coroutine must be awaited before using any other part of the SDK::

            await FoundryLocalManager.initialize(config)
            manager = FoundryLocalManager.instance

        Args:
            config: Configuration object for the SDK.
        """
        # Run the synchronous constructor in a thread to avoid blocking
        # the event loop during native library initialization.
        await asyncio.to_thread(FoundryLocalManager, config)
        
    def __init__(self, config: Configuration):
        # Enforce singleton creation under a class-level lock and ensure
        # that `FoundryLocalManager.instance` is set exactly once.
        with FoundryLocalManager._lock:
            if FoundryLocalManager.instance is not None:
                raise FoundryLocalException(
                    "FoundryLocalManager is a singleton and has already been initialized."
                )
            config.validate()
            self.config = config
            self._initialize()
            FoundryLocalManager.instance = self

        self.urls = None
        self._async_lock = asyncio.Lock()

    def _initialize(self):
        set_default_logger_severity(self.config.log_level)

        external_service_url = self.config.web.external_url if self.config.web else None

        self._core_interop = CoreInterop(self.config)
        self._model_load_manager = ModelLoadManager(self._core_interop, external_service_url)
        self.catalog = Catalog(self._model_load_manager, self._core_interop)

    async def discover_eps(self) -> list[EpInfo]:
        """Discover available execution providers and their registration status.

        Returns:
            List of ``EpInfo`` entries for all discoverable EPs.

        Raises:
            FoundryLocalException: If EP discovery fails or response JSON is invalid.
        """
        response = await self._core_interop.execute_command("discover_eps")
        if response.error is not None:
            raise FoundryLocalException(f"Error discovering execution providers: {response.error}")

        try:
            adapter = TypeAdapter(List[EpInfo])
            return adapter.validate_json(response.data or "[]")
        except Exception as e:
            raise FoundryLocalException(
                f"Failed to decode JSON response from discover_eps: {e}. Response was: {response.data}"
            ) from e

    async def download_and_register_eps(
        self,
        names: Optional[list[str]] = None,
        progress_callback: Optional[Callable[[str, float], None]] = None,
    ) -> EpDownloadResult:
        """Download and register execution providers.

        Args:
            names: Optional subset of EP names to download. If omitted or empty,
                all discoverable EPs are downloaded.
            progress_callback: Optional callback ``(ep_name: str, percent: float) -> None``
                invoked as each EP downloads. ``percent`` is 0-100.

        Returns:
            ``EpDownloadResult`` describing operation status and per-EP outcomes.

        Raises:
            FoundryLocalException: If the operation fails or response JSON is invalid.
        """
        request = None
        if names is not None and len(names) > 0:
            request = InteropRequest(params={"Names": ",".join(names)})

        if progress_callback is not None:
            loop = asyncio.get_running_loop()

            def _on_chunk(chunk: str) -> None:
                sep = chunk.find("|")
                if sep >= 0:
                    ep_name = chunk[:sep] or ""
                    try:
                        percent = float(chunk[sep + 1:])
                        loop.call_soon_threadsafe(progress_callback, ep_name, percent)
                    except ValueError:
                        pass

            response = await self._core_interop.execute_command_with_callback(
                "download_and_register_eps", request, _on_chunk
            )
        else:
            response = await self._core_interop.execute_command("download_and_register_eps", request)

        if response.error is not None:
            raise FoundryLocalException(f"Error downloading execution providers: {response.error}")

        if response.data:
            try:
                adapter = TypeAdapter(EpDownloadResult)
                ep_result = adapter.validate_json(response.data)
            except Exception as e:
                raise FoundryLocalException(
                    "Failed to decode JSON response from download_and_register_eps: "
                    f"{e}. Response was: {response.data}"
                ) from e
        else:
            ep_result = EpDownloadResult(
                Success=True, Status="Completed", RegisteredEps=[], FailedEps=[]
            )

        # Invalidate the catalog cache if any EP was newly registered so the next access
        # re-fetches models with the updated set of available EPs.
        if ep_result.success or len(ep_result.registered_eps) > 0:
            await self.catalog._invalidate_cache()

        return ep_result

    async def start_web_service(self):
        """Start the optional web service.

        If provided, the service will be bound to the value of Configuration.web.urls.
        The default of http://127.0.0.1:0 will be used otherwise, which binds to a random ephemeral port.

        FoundryLocalManager.urls will be updated with the actual URL/s the service is listening on.
        """
        async with self._async_lock:
            response = await self._core_interop.execute_command("start_service")

            if response.error is not None:
                raise FoundryLocalException(f"Error starting web service: {response.error}")

            bound_urls = json.loads(response.data)
            if bound_urls is None or len(bound_urls) == 0:
                raise FoundryLocalException("Failed to get bound URLs from web service start response.")

            self.urls = bound_urls

    async def stop_web_service(self):
        """Stop the optional web service."""

        async with self._async_lock:
            if self.urls is None:
                raise FoundryLocalException("Web service is not running.")

            response = await self._core_interop.execute_command("stop_service")

            if response.error is not None:
                raise FoundryLocalException(f"Error stopping web service: {response.error}")

            self.urls = None
